jdaviz/configs/default/plugins/plot_options/plot_options.py

```diff
@@ -753,135 +767,81 @@ def _update_stretch_histogram_colorbar(self, msg={}):
-             'stretch_hist_nbins', 'image_contrast_value', 'image_bias_value',
-             'stretch_curve_visible')
-    @skip_if_no_updates_since_last_active()
-    def _update_stretch_curve(self, msg=None):
-        mark_label_prefix = "stretch_curve: "
-        knots_label_prefix = "stretch_knots: "
-
-        if not self._viewer_is_image_viewer() or not hasattr(self, 'stretch_histogram'):
-            # don't update histogram if selected viewer is not an image viewer,
-            # or the stretch histogram hasn't been initialized:
-            return
-
-        if not self.stretch_curve_visible:
-            # clear marks if curve is not visible:
-            for existing_mark_label, mark in self.stretch_histogram.marks.items():
-                if (existing_mark_label.startswith(mark_label_prefix) or
-                        existing_mark_label.startswith(knots_label_prefix)):
-                    # clear this mark
-                    mark.x = []
-                    mark.y = []
-            return
-
-        for layer in self.layer.selected_obj:
-            if isinstance(layer.layer, GroupedSubset):
-                # don't update histogram for subsets:
-                continue
-
-            # clear old mark, if it exists:
-            mark_label = f'{mark_label_prefix}{layer.label}'
-            mark_exists = mark_label in self.stretch_histogram.marks
-
-            knot_label = f"{knots_label_prefix}{layer.label}"
-            # create the new/updated mark following the colormapping
-            # procedure in glue's CompositeArray:
-            interval = ManualInterval(self.stretch_vmin.value, self.stretch_vmax.value)
-            contrast_bias = ContrastBiasStretch(layer.state.contrast, layer.state.bias)
-            stretch = stretches.members[layer.state.stretch]
-            layer_cmap = layer.state.cmap
-
-            if isinstance(stretch, SplineStretch):
-                knot_x = (self.stretch_vmin_value +
-                          stretch.x * (self.stretch_vmax_value - self.stretch_vmin_value))
-                knot_y = stretch.y
+        # create the new/updated stretch curve following the colormapping
+        # procedure in glue's CompositeArray:
+        interval = ManualInterval(self.stretch_vmin_value, self.stretch_vmax_value)
+        contrast_bias = ContrastBiasStretch(self.image_contrast_value, self.image_bias_value)
+        stretch = stretches.members[self.stretch_function_value]
+        layer_cmap = layer.state.cmap
+
+        # show the colorbar
+        color_mode = self.image_color_mode_value
+
+        # NOTE: Index 0 in marks is assumed to be the bin centers.
+        x = self.stretch_histogram.figure.marks[0].x
+        y = np.ones_like(x)
+
+        # Copied from the __call__ internals of glue/viewers/image/composite_array.py
+        data = interval(x)
+        data = contrast_bias(data, out=data)
+        data = stretch(data, out=data)

```

Comment:
I wonder if we can consolidate this with the similar logic for the stretch curve... the stretch curve uses a linspace between vmin/vmax whereas this uses the histogram bins and needs to extend beyond vmin/vmax.   (@pllim @bmorris3 - what if we did a linspace between the lower and upper bin and use the same sampling for both?

Reply:
colorbar needs to render beyond the vertical lines, it needs to span the whole visible histogram. I think the lines can move around, but not the histogram bars. Or am I not understanding how all these stuff is supposed to work together?

Reply:
ok, let's merge for now since @haticekaratay and @javerbukh both need to build on top of this and we can discuss and do this as follow-up if we think it can be simplified.  Thanks!